File: tictactoe/tic_tac_toe.rb
class TicTacToe
    attr_reader :state

    def initialize(state)
      @state = state
    end

    def getWinner()
        horizontal_win = win?(state)
        return horizontal_win if [0, 1].include?(horizontal_win)

        vertical_win = win?(state.transpose)
        return vertical_win if [0, 1].include?(vertical_win)

        # diagonal win
        win?([
            [state[0][0], state[1][1], state[2][2]],
            [state[0][2], state[1][1], state[2][0]]
            ])
     end

     def win?(rows)
        rows.each do |row|
            return 0 if row.uniq == [0]
            return 1 if row.uniq == [1]
        end
        nil
     end

end

state = [[0, 0, 0], [1, 0, nil], [1, nil, nil]]

game = TicTacToe.new(state)
winner = game.getWinner()

if winner.nil?
    puts "No winners"
else
    puts "Player #{winner} wins"
end

File: tictactoe/tic_tac_toe.py
class TicTacToe:
    def __init__(self, state):
        self.state = state

    def winner(self):
        for segment in [self.state, zip(*self.state), self.diagonals()]:
            winner = self.check_win(segment)
            if winner in [0, 1]:
                return winner
        return None

    def diagonals(self):
        n = len(self.state)
        main_diagonal = [self.state[i][i] for i in range(n)]
        anti_diagonal = [self.state[i][n - i - 1] for i in range(n)]
        return [main_diagonal, anti_diagonal]

    def check_win(self, segment):
        for row in segment:
            if all(cell == 0 for cell in row):
                return 0
            if all(cell == 1 for cell in row):
                return 1
        return None

state = [[1, 0, 0], [1, 1, None], [1, 0, 1]]
game = TicTacToe(state)
winner = game.winner()

print(f"Player {winner} wins" if winner is not None else "No winner")
